presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java

```diff
@@ -26,17 +26,15 @@ private ParquetTypeUtils()
     public static parquet.schema.Type getParquetType(HiveColumnHandle column, MessageType messageType, boolean useParquetColumnNames)
     {
         if (useParquetColumnNames) {
-            if (messageType.containsField(column.getName())) {
-                return messageType.getType(column.getName());
-            }
-            // parquet is case-sensitive, but hive is not. all hive columns get converted to lowercase
-            // check for direct match above but if no match found, try case-insensitive match
-            for (Type type : messageType.getFields()) {
-                if (type.getName().equalsIgnoreCase(column.getName())) {
-                    return type;
-                }
+            String name = column.getName();
+            Type type = getParquetTypeByName(name, messageType);
+
+            // when a parquet field is a hive keyword we append an _ to it in hive. When doing
```

Comment:
If a field really contains a trailing _, it will be picked up in the direct match check.

I refactored to disrupt as few of the existing lines as possible and add our custom code below. This method gets called for a number of places in the code so I'd like to avoid changing all of those.

The logic is a little more than you describe, which I've added in the javadocs:
- direct match
- ignore case match
- strip _ direct match
- strip _ ignore case match